---
title: "`r paste('Report for ', params$name)`"
author: "Matt Hewitt"
date: "`r format(Sys.time(), '%d-%m-%Y')`"
params:
  name:             name
  pcs:              number_of_pcs
  
  raw_results_file: raw_results
  combat_data_file: combat_data
  pheno_data_file:  pheno_data
  corr_plot_file:   corr_plot
  
  combat_data:  "../" 
  raw_results:  "./results/"
  pheno_data:   "./objects/"
  corr_plot:    "./pipeline_outs/"
  
output: 
  html_document:
    number_sections: true
    fig_caption: yes
    df_print: paged
    #toc: true
    #toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#setwd("E:/BMIQ MAV/19-07-30-split_cohort_and_run_season/reports/tG_summer")
knitr::opts_knit$set(root.dir = here::here())
getwd()
```

```{r libraries, include=FALSE}
library(clipr)
library(GenABEL)
library(ggplot2)
library(missMethyl)
library(limma) 
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(DMRcate)
library(tidyverse)
library(kableExtra)
library(formattable)
library(Hmisc)
```

```{r set-up directories, include=FALSE}

dir.create(paste0("./reports", "/", params$name))

data_output <-      paste0("./reports", "/", params$name, "/")

raw_data <-         paste0(params$raw_results, params$raw_results_file)
combat_data_raw <-  readRDS(paste0(params$combat_data, params$combat_data_file)) 
pheno_data_raw <-   readRDS(paste0(params$pheno_data, params$pheno_data_file))         
corrplot <-         read.table(paste0(params$corr_plot, params$corr_plot_file), 
                               sep = "\t", 
                               header=T
                               )

###### ALSO FILL IN SvA PLOT PATH BELOW #######

cols<-c(
  "Row.names",
  "logFC",
  "AveExpr",
  "t",
  "P.Value",
  "adj.P.Val",
  "B",
  "chr",
  "pos",
  "strand",
  "Name",
  "Relation_to_Island",
  "UCSC_RefGene_Name",
  "UCSC_RefGene_Accession",
  "UCSC_RefGene_Group",
  "Regulatory_Feature_Group"
)

var_raw<-read_tsv(paste0(raw_data), 
         col_names = F,
         skip = 1, 
         na = character()
         )
var<-var_raw[,-1] 
colnames(var)<-cols
var <- var %>% arrange(adj.P.Val) %>% rowid_to_column("Rank")
var$UCSC_RefGene_Name<-sub(";.*", "", var$UCSC_RefGene_Name)
var$UCSC_RefGene_Group<-sub(";.*", "", var$UCSC_RefGene_Group)
```

<br />
<br />

# Top 50 hits

```{r top 50 genes, echo=FALSE, warning=FALSE}
var %>% 
  mutate_at(7, funs(round(., 3))) %>% 
  arrange(adj.P.Val) %>% 
  dplyr::slice(1:50) %>% 
  mutate(adj.P.Val = cell_spec(
    adj.P.Val,
    color = ifelse(adj.P.Val < 0.25, "white", "black"), 
    bold = ifelse(adj.P.Val < 0.25, T, F),
    background = ifelse(adj.P.Val < 0.25, "green", "white"),
  )) %>%
  kable(
   booktabs = TRUE,
   linesep = "",
   align = "c",
   caption = " ",
   escape = F) %>%
   kable_styling(
      bootstrap_options = c("striped", 
                            "hover", 
                            "condensed",
                            full_width = F,
                            position = "center",
                            fixed_thead = T)
   ) %>% 
    scroll_box(width = "100%", height = "600px")

```

```{r plotted tophits}

```

<br />
<br />

# Top 100 {.tabset .tabset-fade}

## Top 100 named genes

```{r top 100 named genes, include=FALSE}
named_hits <- var[!(is.na(var$UCSC_RefGene_Name) | var$UCSC_RefGene_Name == ""),]
top100_named_genes <-
  named_hits %>%
  arrange(adj.P.Val) %>%
  slice(1:100) %>%
  dplyr::select(UCSC_RefGene_Name) %>% 
  readr::write_csv(paste0(data_output,"top100_named_genes.csv"), col_names = T)
```

```{r top 100 named genes graph, echo=FALSE}
top100_named_genes %>%
  kable(
   booktabs = TRUE,
   linesep = "",
   align = "c",
   caption = " ") %>%
   kable_styling(
      bootstrap_options = c("striped", 
                            "hover", 
                            "condensed",
                            full_width = F,
                            position = "center",
                            fixed_thead = T)
   ) %>% 
    scroll_box(height = "400px")
```

<br />
<br />

## Top 100 cgs from named hits

```{r top 100 cgs, include=FALSE}
top100_cgs <- named_hits %>%
  arrange(adj.P.Val) %>%
  slice(1:100) %>%
  dplyr::select(Row.names) %>% 
  readr::write_csv(paste0(data_output,"top100_cgs.csv"), col_names = T)
```

```{r top 100 cgs graph, echo=FALSE}
top100_cgs %>%
  kable(
   booktabs = TRUE,
   linesep = "",
   align = "c",
   caption = " ") %>%
   kable_styling(
      bootstrap_options = c("striped", 
                            "hover", 
                            "condensed",
                            full_width = F,
                            position = "center",
                            fixed_thead = T)
   ) %>% 
    scroll_box(height = "400px")
```

<br />
<br />

## Significant named genes

```{r top sig genenames, include=FALSE}
sig_genenames <-
  var %>%
  arrange(adj.P.Val) %>%
  filter(adj.P.Val <0.25) %>%
  dplyr::select(UCSC_RefGene_Name) %>% 
  readr::write_csv(paste0(data_output,"sig_genenames.csv"), col_names = T)
```

```{r top sig genenames table, echo=FALSE}
sig_genenames %>%
  kable(
   booktabs = TRUE,
   linesep = "",
   align = "c",
   caption = " ") %>%
   kable_styling(
      bootstrap_options = c("striped", 
                            "hover", 
                            "condensed",
                            full_width = F,
                            position = "center",
                            fixed_thead = T)
   ) %>% 
    scroll_box(height = "400px")
```

<br />
<br />

## Significant cgs from named hits

```{r top sig cgs, include=FALSE}
sig_cgs <-
  var %>%
  arrange(adj.P.Val) %>%
  filter(adj.P.Val <0.25) %>%
  dplyr::select(Row.names) %>% 
  readr::write_csv(paste0(data_output,"top100_cgs.csv"), col_names = T)
```

```{r top sig cgs graph, echo=FALSE}
sig_cgs %>%
  kable(
   booktabs = TRUE,
   linesep = "",
   align = "c",
   caption = " ") %>%
   kable_styling(
      bootstrap_options = c("striped", 
                            "hover", 
                            "condensed",
                            full_width = F,
                            position = "center",
                            fixed_thead = T)
   ) %>% 
    scroll_box(height = "400px")
```

# Lambdas and qqplot {.tabset .tabset-fade}

## Lambdas

```{r sums and lambdas, echo=FALSE}

result_list <- data.frame(
  "Lambda" = estlambda(var$P.Value, method = "median"),
  "FDR<0.25" = sum(var$adj.P.Val < 0.25),
  "FDR<0.20" = sum(var$adj.P.Val < 0.20),
  "FDR<0.15" = sum(var$adj.P.Val < 0.15),
  "FDR<0.10" = sum(var$adj.P.Val < 0.10),
  "FDR<0.05" = sum(var$adj.P.Val < 0.05)
)
row.names(result_list)<-params$pcs

result_list %>%
  kable(
   booktabs = TRUE,
   linesep = "",
   align = "c",
   caption = " ") %>%
   kable_styling(
      bootstrap_options = c("striped", 
                            "hover", 
                            "condensed",
                            full_width = F,
                            position = "center",
                            fixed_thead = T)
   ) 
```

<br />
<br />

## qqplot and nominal p.values

```{r qqplot function, include=FALSE}

var_ps <- var$P.Value

gg_qqplot <- function(var_ps, ci = 0.95) {
  n  <- length(var_ps)
  df <- data.frame(
    observed = -log10(sort(var_ps)),
    expected = -log10(ppoints(n)),
    clower   = -log10(qbeta(p = (1 - ci) / 2, shape1 = 1:n, shape2 = n:1)),
    cupper   = -log10(qbeta(p = (1 + ci) / 2, shape1 = 1:n, shape2 = n:1))
  )
  log10Pe <- expression(paste("Expected -log"[10], plain(P)))
  log10Po <- expression(paste("Observed -log"[10], plain(P)))
  ggplot(df) +
    geom_point(aes(expected, observed), shape = 1, size = 3) +
    geom_abline(intercept = 0, slope = 1, alpha = 0.5) +
    geom_line(aes(expected, cupper), linetype = 2) +
    geom_line(aes(expected, clower), linetype = 2) +
    xlab(log10Pe) +
    ylab(log10Po) +
    ggtitle("qqplot")
}
```

```{r qqplot and nom. p value histogram, echo=FALSE}

gg_qqplot(var_ps)
ggplot(var, aes(x=P.Value)) + 
  geom_histogram() +
  ggtitle("P-values: var")
```

<br />
<br />

## Correlation plot

```{r correlation plot, echo=FALSE}

#corrplot <- read.table("../../pipeline_outs/corr_plot_area.txt", sep = "\t", header=T)

#corrplot <- as.data.frame(format.df(corrplot, numeric.dollar=FALSE, scientific = -4,4))

 corrplot %>%
   rownames_to_column() %>%
  mutate(
    PC1 = cell_spec(PC1, "html", background = ifelse(PC1 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC2 = cell_spec(PC2, "html", background = ifelse(PC2 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC3 = cell_spec(PC3, "html", background = ifelse(PC3 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC4 = cell_spec(PC4, "html", background = ifelse(PC4 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC5 = cell_spec(PC5, "html", background = ifelse(PC5 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC6 = cell_spec(PC6, "html", background = ifelse(PC6 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC7 = cell_spec(PC7, "html", background = ifelse(PC7 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC8 = cell_spec(PC8, "html", background = ifelse(PC8 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC9 = cell_spec(PC9, "html", background = ifelse(PC9 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC10 = cell_spec(PC10, "html", background = ifelse(PC10 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC11 = cell_spec(PC11, "html", background = ifelse(PC11 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC12 = cell_spec(PC12, "html", background = ifelse(PC12 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC13 = cell_spec(PC13, "html", background = ifelse(PC13 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC14 = cell_spec(PC14, "html", background = ifelse(PC14 < 0.05, "#f0c2c2", "#c3f0c2")),
    PC15 = cell_spec(PC15, "html", background = ifelse(PC15 < 0.05, "#f0c2c2", "#c3f0c2"))
    ) %>%
  kable(
    booktabs = TRUE,
    linesep = "",
   align = c("l", rep("c", 15)),
   caption = " ",
   escape = F) %>%
   kable_styling(
      bootstrap_options = c("striped",
                            "hover",
                            "condensed",
                            full_width = F,
                            position = "center",
                            fixed_thead = T)
   ) 
   # scroll_box(width = "150%", height = "600px")



```

<br />
<br />

## Other plots

[SVA plots](./pipeline_outs/Rplots.pdf)

# Pie chart data {.tabset .tabset-fade}

```{r pie chart data, include=FALSE}

## Relation to island 
pie_rel_to_island_all <-
  var %>% 
  group_by(Relation_to_Island) %>% 
  tally() %>%  
  write_csv(paste0(data_output,"pie_chart_data.csv"), col_names = T)

pie_region_100 <-
  var[!(is.na(var$Relation_to_Island) | var$Relation_to_Island == ""),]
pie_rel_to_island_100 <-
  pie_region_100 %>%
  arrange(adj.P.Val) %>%
  dplyr::slice(1:100) %>%
  dplyr::select(Relation_to_Island) %>%
  group_by(Relation_to_Island) %>%
  tally() %>% 
  #rename(Relation_to_Island_100 = Relation_to_Island) %>% 
  readr::write_csv(paste0(data_output, "pie_chart_data.csv"), col_names = T, append = T)

## RefGene group
pie_ref_group_all <-
  var %>% 
  group_by(UCSC_RefGene_Group) %>% 
  tally() %>% 
  write_csv(paste0(data_output, "pie_chart_data.csv"), col_names = T, append = T)

pie_region_100 <-
  var[!(is.na(var$UCSC_RefGene_Group) | var$UCSC_RefGene_Group == ""),]
pie_ref_group_100 <-
  pie_region_100 %>%
  arrange(adj.P.Val) %>%
  dplyr::slice(1:100) %>%
  dplyr::select(UCSC_RefGene_Group) %>%
  group_by(UCSC_RefGene_Group) %>%
  tally() %>% 
  #rename(UCSC_RefGene_Group_100 = UCSC_RefGene_Group) %>% 
  readr::write_csv(paste0(data_output, "pie_chart_data.csv"), col_names = T, append = T)
```

<br />
<br />

## Relationship to island - all cpgs

```{r pie chart tables, echo=FALSE}

kable(
  pie_rel_to_island_all,
  "html",
  booktabs = TRUE,
  linesep = "",
  align = "c",
  caption = " ",
  escape = F
) %>%
  kable_styling(
    bootstrap_options = c(
      "striped",
      "hover",
      "condensed",
      full_width = F,
      position = "center",
      fixed_thead = T
    )
  )
```

<br />
<br />

## Relationship to island - top 100 cpgs

```{r, echo=FALSE}
pie_rel_to_island_100 %>% 
kable(
  booktabs = TRUE,
  linesep = "",
  align = "c",
  caption = " ",
  escape = F) %>%
  kable_styling(
    bootstrap_options = c(
      "striped",
      "hover",
      "condensed",
      full_width = F,
      position = "center",
      fixed_thead = T
    )
  )
```

<br />
<br />

## Refgene group - all cpgs

```{r, echo=FALSE}
pie_ref_group_all %>% 
kable(
  booktabs = TRUE,
  linesep = "",
  align = "c",
  caption = " ",
  escape = F) %>%
  kable_styling(
    bootstrap_options = c(
      "striped",
      "hover",
      "condensed",
      full_width = F,
      position = "center",
      fixed_thead = T
    )
  )
```

<br />
<br />

## Refgene group - all cpgs

```{r, echo=FALSE}
pie_ref_group_100 %>% 
kable(
  booktabs = TRUE,
  linesep = "",
  align = "c",
  caption = " ",
  escape = F) %>%
  kable_styling(
    bootstrap_options = c(
      "striped",
      "hover",
      "condensed",
      full_width = F,
      position = "center",
      fixed_thead = T
    )
  ) 
```

<br />
<br />

# missMethyl {.tabset .tabset-fade}

```{r missMethyl, include = FALSE}
# missMethyl ---------------------------------------------------------------------------------------------------

Universe<-var$Row.names
Sig_0.25<-var$Row.names[var$adj.P.Val<0.25]
top100_named_genes <- var[!(is.na(var$UCSC_RefGene_Name) | var$UCSC_RefGene_Name == ""),]
top_100_cpgs <- top100_named_genes %>% arrange(adj.P.Val) %>%  dplyr::slice(1:100) %>% pull(Row.names)

# FDR<0.25
# Sig_0.25<-as.character(Sig_0.25)
# Universe<-as.character(Universe)
# gst_0.25_GO<-gometh(sig.cpg=Sig_0.25,
#                      all.cpg = Universe,
#                      collection = "GO",
#                      array.type = "EPIC"
#                      )
# colnames(gst_0.25_GO)[2]<-"Term"
# 
# ## Results 0.25
# Results_0.25_GO <- topGO(gst_0.25_GO, number = 50)
# Results_0.25_GO <- Results_0.25_GO %>%
#    arrange(ONTOLOGY) %>%
#    readr::write_csv(paste0(data_output, "./missMethyl_results.csv"), col_names = T)

## 1:100
#top_100_cpgs<-as.character(top_100_cpgs) Tests gene ontology enrichment for
#significant CpGs from Illumina's Infinium HumanMethylation450 or
#MethylationEPIC array, taking into account the differing number of probes per
#gene present on the array.
Universe<-as.character(Universe)
gst_top100_go<-gometh(sig.cpg=top_100_cpgs, 
                      all.cpg = Universe, 
                      collection = "GO", 
                      array.type = "EPIC"
                      )
colnames(gst_top100_go)[2]<-"Term"

### Results 1:100 - topGO pulls the most significant from the pack
gst_top100_go<-
results_top100_go <- topGO(gst_top100_go, number = 50)
results_top100_go <- 
  results_top100_go %>% 
  arrange(ONTOLOGY) %>% 
  readr::write_csv(paste0(data_output, "missMethyl_results.csv"), col_names = T, append = T)
```

## Significant hits (FDR<0.25)

```{r miss methyl tables, echo = FALSE}
# Results_0.25_GO %>%
# kable(
#   booktabs = TRUE,
#   linesep = "",
#   align = c("c", "l", "c", "c", "c", "c"),
#   caption = " ",
#   escape = F
# ) %>%
#   kable_styling(
#     bootstrap_options = c(
#       "striped",
#       "hover",
#       "condensed",
#       full_width = F,
#       position = "center",
#       fixed_thead = T
#     )
#   )

```

## Top 100 cpgs

```{r, echo = FALSE}
kable(
  results_top100_go,
  "html",
  booktabs = TRUE,
  linesep = "",
  align = c("c", "l", "c", "c", "c", "c"),
  caption = " ",
  escape = F
) %>%
  kable_styling(
    bootstrap_options = c(
      "striped",
      "hover",
      "condensed",
      full_width = F,
      position = "center",
      fixed_thead = T
    )
  ) 
```

<!-- # DMRcate ------------------------------------------------------------------------------------------------------ -->

<!-- ## read-in data -->

<!-- # samplesheet <- samplesheet_raw -->
<!-- # combat_data <- combat_data_raw -->
<!-- # pheno_data <- pheno_data_raw -->
<!-- #  -->
<!-- #  -->
<!-- # #Change 'beta' in the next step to the name of your beta table variable -->
<!-- # pheno_data<-pheno_data[!is.na(pheno_data$etotbmc), ] -->
<!-- #  -->
<!-- # samplenames<-as.character(pheno_data$Sample_Name) -->
<!-- # samplenames<-as.data.frame(samplenames) -->
<!-- # subset<-combat_data[ ,samplenames$samplenames] -->
<!-- #  -->
<!-- # mod <- model.matrix(~etotbmc, data=pheno_data) -->
<!-- #  -->
<!-- # set.seed(12345) -->
<!-- # myAnnotation <- cpg.annotate(datatype="array",  -->
<!-- #                              subset,  -->
<!-- #                              analysis.type="differential",  -->
<!-- #                              what="Beta",  -->
<!-- #                              arraytype="EPIC",  -->
<!-- #                              design=mod,  -->
<!-- #                              contrasts=FALSE,  -->
<!-- #                              coef=2,  -->
<!-- #                              fdr=0.25 -->
<!-- # ) -->
<!-- #  -->
<!-- # set.seed(12345) -->
<!-- # dmroutput <- dmrcate(myAnnotation, lambda=1000, C=2) -->
<!-- # results.ranges <- extractRanges(dmroutput, genome="hg19") -->


